<a href="https://colab.research.google.com/github/jorgepdsML/ARTIFICIAL_INTELLIGENCE_TENSORFLOW/blob/master/UNIDADIII_OLS_ESTIMACION_2021.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### PREGUNTA N°2 DEL EXAMEN DE PROCESOS ESTOCASTICOS
##### ALUMNO: Jorge Miranda

In [1]:
#Establecer el backend de matplotlib en el backend 'en línea' :
%matplotlib inline 


In [2]:
"""
--numpy: Biblioteca que da soporte a la creación y manipulación
de matrices multidimensionales asi como funciones matematicas de alto nivel.
--importar numpy como np
"""
import numpy as np
#Apartir de aqui se usara np en lugar de numpy
"""
--pyplot permite realizar distintos estilos de ploteos 
con el fin de realizar un analisis grafico sobre los datos.
--importar el modulo pyplot del paquete matplotlib como plt
"""
import matplotlib.pyplot as plt
#Apartir de aqui se usara plt en lugar de pyplot 
"""
--importar api del paquete statsmodels como sm
--stastsmodels: Paquete de python , complemento de scipy, orientado
a la estadistica.
--api:modulo que brinda modelos y métodos estadisticos.
"""
import statsmodels.api as sm
#A partir de aqui se usara sm en lugar de api 
from statsmodels.sandbox.regression.predstd import wls_prediction_std


/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [3]:
#Establecer la semilla para el algoritmo generador de números pseudoaleatorios en PythonE
np.random.seed(9876789)

### Creación de datos artificiales
Se considera el uso de la generación datos artificiales para validar el uso de la estimación OLS    \
Se considera en generar 100 datos de la variable de regresión $x(t)$ con el fin de explicar $y(t)$ empleando un modelo de regresión cuadratico \
$nsample=100$ 

| $x(t)$ | $y(t)$ |
| :---: | :---:|
| $x(0)$| $y(0)$|
| $x(1)$| $y(2)$|
| $x(2)$| $y(2)$|
| $x(3)$| $y(3)$|
| $.$| $.$|
| $.$| $.$|
| $.$| $.$|
| $.$| $.$|
| $x(nsample-1)$| $y(nsample-1)$|

### Modelo de Regresión  
El modelo de regresión viene definido por la siguiente ecuación    \
$ y(t)=\beta_{0}+\beta_{1}x(t) +\beta_{2}x^{2}(t)+\xi(t) $

Donde los datos $x(t), x(t)^2$ son generados artificialmente

El modelo de regresión ahora se define como:

$ y(t)=\beta_{0}+\beta_{1}x_1(t) +\beta_{2}x_2(t)+\xi(t) $


In [4]:
#Definir la cantidad de muestras
nsample = 100
"""
Definir un vector de 100 elementos 
con valores desde 0 hasta 10
"""
x = np.linspace(0, 10, 100)
#Mostrar vectot creado
print("---------------Vector x---------------- \r\n",x)
print(x)
"""
Realizar el apilamiento por columna al vector x con el 
vector x**2 en la cual se crea una matriz 2D y se asigna
a X.
"""
X = np.column_stack((x, x**2))
#Mostrar la matriz creada
print("\r\n---------------Matriz X---------------- \r\n",X)
"""
Definir un vector de 3 elementos
que consisten en 1 , 0.1 y 10
"""
beta = np.array([1, 0.1, 10])
#Mostrar el vector beta
print("\r\n---------------Vector beta---------------- \r\n",beta)
"""
Crear un vector aleatorio con distribución  
normal con una cantidad de muestras igual a nsample
"""
e = np.random.normal(size=nsample)
#Mostrar el vector de error
print("\r\n---------------Vector de error---------------- \r\n",e)


---------------Vector x---------------- 
 [ 0.          0.1010101   0.2020202   0.3030303   0.4040404   0.50505051
  0.60606061  0.70707071  0.80808081  0.90909091  1.01010101  1.11111111
  1.21212121  1.31313131  1.41414141  1.51515152  1.61616162  1.71717172
  1.81818182  1.91919192  2.02020202  2.12121212  2.22222222  2.32323232
  2.42424242  2.52525253  2.62626263  2.72727273  2.82828283  2.92929293
  3.03030303  3.13131313  3.23232323  3.33333333  3.43434343  3.53535354
  3.63636364  3.73737374  3.83838384  3.93939394  4.04040404  4.14141414
  4.24242424  4.34343434  4.44444444  4.54545455  4.64646465  4.74747475
  4.84848485  4.94949495  5.05050505  5.15151515  5.25252525  5.35353535
  5.45454545  5.55555556  5.65656566  5.75757576  5.85858586  5.95959596
  6.06060606  6.16161616  6.26262626  6.36363636  6.46464646  6.56565657
  6.66666667  6.76767677  6.86868687  6.96969697  7.07070707  7.17171717
  7.27272727  7.37373737  7.47474747  7.57575758  7.67676768  7.77777778
  7.87878

### Notación matricial
El modelo de regresión es descrito matricialmente como   \
$ Y=XB+E $  

$Y=[y(0),y(1),..... y(99)]^T$  

$B=[1,0.1,10]^T$  


$X=
 \begin{bmatrix}
1 & x_1(0) & x_2(0)\\
1 & x_1(1) & x_2(1)\\
1 & x_1(2) & x_2(2)\\
. & . &-\\
. & . & \\
1 & x_1(99) & x_2(99)\\
\end{bmatrix}_{100x3}$  
 
$E=[\xi(0),\xi(1),.....\xi(99)]^T$
 

In [ ]:
"""
Emplear la función add_constant para
agregar una columna de unos a la matriz X.
"""
X = sm.add_constant(X)
"""
Realizar el producto entre X y el vector beta y sumarle
el vector residuo de predicción
"""
y = np.dot(X, beta) + e
print("---------------Vector y---------------- \r\n",y)


---------------Vector y---------------- 
 [1.59484106e-01 2.70962667e+00 1.89386568e+00 1.44041774e+00
 5.60719622e+00 4.01415960e+00 3.79042816e+00 5.49896267e+00
 7.93934063e+00 9.31871940e+00 1.16151797e+01 1.30800908e+01
 1.65916731e+01 1.90612419e+01 2.22779839e+01 2.22973772e+01
 2.90728303e+01 2.95174936e+01 3.43906010e+01 3.68997190e+01
 4.15314299e+01 4.65124798e+01 4.87622029e+01 5.54649378e+01
 6.03995721e+01 6.68905550e+01 7.10273234e+01 7.57840504e+01
 8.32307851e+01 8.80067872e+01 9.24604592e+01 1.00869036e+02
 1.05000770e+02 1.13078878e+02 1.18651097e+02 1.26357451e+02
 1.32682080e+02 1.39131512e+02 1.47437302e+02 1.56109386e+02
 1.64097051e+02 1.73688174e+02 1.81821259e+02 1.88890813e+02
 1.99352099e+02 2.10227715e+02 2.16641136e+02 2.28077202e+02
 2.37086627e+02 2.45435044e+02 2.56253216e+02 2.66268455e+02
 2.77388427e+02 2.88835852e+02 2.97996070e+02 3.11225764e+02
 3.20222632e+02 3.34500736e+02 3.44501923e+02 3.54912182e+02
 3.69614763e+02 3.79619250e+02 3.91610439e+

### Estimación por Minimos Cuadrados Ordinarios

Dado las observaciones $y(t)$ y la matriz $X$ se debe de estimar los valores $\beta_{0}$, $\beta_{1}$ y  $\beta_{2}$ tal que se minimice la suma de errores cuadraticos.  

In [ ]:
"""
Crear un modelo de regresión tomandos los datos de 
la matriz X y del vector y"""
model = sm.OLS(y, X)
"""
Emplear el método fit() para realizar el ajuste del modelo
con los datos reales.Aqui se estima los parametros beta
 """
results = model.fit()
"""
Mostrar un resumen de los resultados obtenidos.
Con este resumen se puede analizar la importancia estadistica 
de las variables que fueron utilizados para explicar y(t).
"""
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                  1.000
Method:                 Least Squares   F-statistic:                 4.020e+06
Date:                Tue, 17 Aug 2021   Prob (F-statistic):          2.83e-239
Time:                        18:21:31   Log-Likelihood:                -146.51
No. Observations:                 100   AIC:                             299.0
Df Residuals:                      97   BIC:                             306.8
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.3423      0.313      4.292      0.0

In [ ]:
""" Mostrar los parametros estimados"""
print("-------------Parametros estimados -----------\r\n",results.params)


-------------Parametros estimados -----------
 [ 1.34233516 -0.04024948 10.01025357]


 ### Analisis
 Los parametros estimados mediante el método OLS son:
 
 
 $\beta_{0}=1.34233516$
 
 $\beta_{1}=-0.04024948$
 
 $\beta_{1}=10.01025357$ 
 
$ y(t)=\beta_{0}+\beta_{1}x_1(t) +\beta_{2}x_2(t)+\xi(t) $
####  VARIABLE ESTADISTICAMENTE NO SIGNIFICATIVA
El coeficiente  $\beta_{1}=-0.04024948 $ implica que la variable $x(t)$
afecta de manera negativa a la variable $y(t)$ pero que se puede
considerar a x(t) como una variable   ESTADISTICAMENTE NO SIGNIFICATIVA, debido a que  
 
$ P(P>|t|) > 5\% $
  

####  VARIABLES ESTADISTICAMENTE  SIGNIFICATIVAS
 
Por otro lado , la variable $x_2(t)$  afecta de manera positiva a la variable $y(t)$ y representa una variable ESTADISTICAMENTE SIGNIFICATIVA asi como el termino constante $\beta_{0}$, debido a que

$ P(P>|t|) < 5\% $